### DISCUSSION TITLE
How fast does the pico boot up?

### DISCUSSION THREAD

**Discussion by drakejest** [earlephilhower/arduino-pico#1303] [Q&A] [answered]
Hello i was wondering how long does the pico reach the setup () portion of the code? I have some time sensitive application that need to get running as fast as possible. I need my pico to to output immediately output a specific PWM signal within 100ms upon receiving power.

Is it achievable ? if not is there a way to achieve it that still uses this core? I would prefer to still remain in the arduino platform :)

I have not timed my pico bootup yet, I will still need to borrow equipment for that . If you have tried timing a pico boot up please share me your results .

**Comment by earlephilhower** [comment, marked as answer]
Sorry, no idea really.  If this is safety critical I really recommend you go bare metal SDK so you can be assured of exactly what is running on the core.  Here, the ROM will boot, then call the OTA stub, which will then call the main Arduino app where the core's own `main` will do a bunch of setup (USB, etc.) before calling any of your own code.

**Reply by drakejest** [reply]
Did a scope test and it just barely passed the 100ms mark. While this is passable, would it be worth exploring skipping analog Freq,Res, and Write. and just go directly with manipulating the registry? I know that a typical arduino board api put lots of overhead checks for their function. Does that exist in this core too?  i might be able to shave down a couple of microseconds

![image](https://user-images.githubusercontent.com/62618765/225092793-f00e25f5-59fc-4668-a73b-f453f9026c2b.png)


```
void setup1() {

  pinMode(1, OUTPUT);
  analogWriteFreq(100000);
  analogWriteResolution(10);
  analogWrite(1, 512);

}

void loop1() {
}

void setup() {
  pinMode(0, OUTPUT);
  analogWriteFreq(100000);
  analogWriteResolution(10);
  analogWrite(0, 512);
}

void loop() {
}
```

**Reply by earlephilhower** [reply]
Any overhead checks for those calls will not add up to 8ms at 133MHZ.  So, I think you need to run bare metal, sorry.  It's obviously right on the edge now and I'm not able to guarantee startup will not get any slower.

**Comment by XZCE** [comment]
I've been tackling this issue too, and I've got it down to 23ms on a real PicoW.

First, I patched boot2 for this specific flash chip to change the ROSC to go faster: https://github.com/raspberrypi/pico-sdk/issues/2281#issuecomment-2771855579
(See the last post from yesterday, it got an extra 20ms, but I still had 67ms of do-nothing time which I started to look into.)

Then I moved the Arduino OTA stuff from here below .text in the linker script (it initialises XOSC with a 64ms delay).  Yeah, I don't use OTA, and a recompile wasn't working for me for some reason, so I just shifted it out of the way to test.

Then I recompiled the libpico stuff from here (which does work for me) with a 6ms XOSC delay (like the SDK does, which is a recent increase, for their choice of crystal).

I do feel like we are being punished with the generic 64ms startup delay, to make this core work with everything.  But I've got a huge list of changes needed for my project "to work" already, so one more makes no real difference.  I just tell people who try to follow my path: tough, it is what it is, use the firmware files I post.  Sad.
